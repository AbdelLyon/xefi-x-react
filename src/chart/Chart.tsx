import type { JSX } from "react";
import React, { forwardRef } from "react";
import type {
  ChartData,
  ChartOptions,
  ChartTypeRegistry,
  InteractionItem,
  TooltipItem,
  Chart as ChartJS,
} from "chart.js";
import {
  CategoryScale,
  Chart as ChartLibrary,
  Legend,
  LinearScale,
  Title,
  Tooltip,
  BarController,
  DoughnutController,
  BarElement,
  ArcElement,
  PointElement,
  LineElement,
  RadialLinearScale,
  ScatterController,
  PolarAreaController,
} from "chart.js";
import type { ChartProps } from "react-chartjs-2";
import { Chart as ChartRoot, getElementAtEvent } from "react-chartjs-2";
import { mergeTailwindClasses } from "@/utils";

ChartLibrary.register(
  CategoryScale,
  LinearScale,
  Title,
  Tooltip,
  Legend,
  RadialLinearScale,
  BarElement,
  ArcElement,
  PointElement,
  LineElement,
  BarController,
  DoughnutController,
  ScatterController,
  PolarAreaController,
);

type ChartType = keyof ChartTypeRegistry;

interface ChartClassNames {
  root?: string;
  canvas?: string;
  title?: string;
  legend?: string;
  tooltip?: string;
}

interface ChartBaseProps<T extends ChartType> {
  data: ChartData<T>;
  options?: ChartOptions<T>;
  type: T;
  getElementSelected?: (elementSelected: InteractionItem[]) => void;
  classNames?: ChartClassNames;
  width?: number | string;
  height?: number | string;
  responsive?: boolean;
  maintainAspectRatio?: boolean;
  title?: string;
  showLegend?: boolean;
  showTooltip?: boolean;
  legendPosition?: "top" | "bottom" | "left" | "right";
  customTooltip?: (context: TooltipItem<T>) => string | string[] | undefined;
}

type Props<T extends ChartType> = ChartBaseProps<T> &
  Omit<ChartProps<T>, keyof ChartBaseProps<T>>;

const defaultClassNames: Required<ChartClassNames> = {
  root: "relative w-full h-max flex flex-col items-center border border-border justify-center bg-white dark:bg-content1 p-6 shadow-md rounded-xl",
  canvas: "w-full h-[400px]",
  title: "text-lg font-semibold text-center mb-4",
  legend: "mt-4",
  tooltip: "bg-white p-2 rounded shadow-lg border text-sm",
};

export const Chart = forwardRef<HTMLDivElement, Props<ChartType>>(
  (
    {
      type,
      data,
      options,
      getElementSelected,
      classNames = {},
      responsive = true,
      maintainAspectRatio = false,
      title,
      showLegend = true,
      showTooltip = true,
      legendPosition = "top",
      customTooltip,
      ...props
    },
    ref,
  ): JSX.Element => {
    // Fusion des classes
    const mergedClassNames = {
      root: mergeTailwindClasses(defaultClassNames.root, classNames.root),
      canvas: mergeTailwindClasses(defaultClassNames.canvas, classNames.canvas),
      title: mergeTailwindClasses(defaultClassNames.title, classNames.title),
      legend: mergeTailwindClasses(defaultClassNames.legend, classNames.legend),
      tooltip: mergeTailwindClasses(
        defaultClassNames.tooltip,
        classNames.tooltip,
      ),
    };

    const handleClick = (event: React.MouseEvent<HTMLCanvasElement>): void => {
      const chartElement = event.currentTarget;
      if (getElementSelected) {
        const clickedElements = getElementAtEvent(
          chartElement as unknown as ChartJS<keyof ChartTypeRegistry>,
          event,
        );

        if (clickedElements.length > 0) {
          getElementSelected(clickedElements);
        }
      }
    };

    const defaultOptions: ChartOptions<ChartType> = {
      responsive,
      maintainAspectRatio,
      plugins: {
        title: title
          ? {
              display: true,
              text: title,
              font: {
                size: 16,
                weight: "bold",
              },
              padding: {
                top: 10,
                bottom: 20,
              },
            }
          : undefined,

        legend: {
          display: showLegend,
          position: legendPosition,
        },

        tooltip: showTooltip
          ? {
              enabled: true,
              backgroundColor: "white",
              titleColor: "#1f2937",
              bodyColor: "#4b5563",
              borderColor: "#e5e7eb",
              borderWidth: 1,
              padding: 8,
              cornerRadius: 4,
              bodyFont: { size: 14 },
              titleFont: {
                size: 14,
                weight: "bold" as const,
              },
              ...(customTooltip && {
                callbacks: {
                  label: customTooltip,
                },
              }),
            }
          : undefined,
      },
    };

    const mergedOptions: ChartOptions<ChartType> = {
      ...defaultOptions,
      ...options,
    };

    return (
      <div ref={ref} className={mergedClassNames.root}>
        <ChartRoot
          data={data}
          options={mergedOptions}
          type={type}
          onClick={handleClick}
          className={mergedClassNames.canvas}
          {...props}
        />
      </div>
    );
  },
);

export type {
  ChartType,
  ChartClassNames,
  ChartTypeRegistry,
  ChartJS,
  ChartOptions,
  ChartProps,
  ChartBaseProps,
};
