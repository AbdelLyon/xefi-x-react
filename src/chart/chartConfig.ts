import {
  CategoryScale,
  Chart as ChartLibrary,
  Legend,
  LinearScale,
  Title,
  Tooltip,
  BarController,
  DoughnutController,
  BarElement,
  ArcElement,
  PointElement,
  LineElement,
  RadialLinearScale,
  ScatterController,
  PolarAreaController,
} from "chart.js";

/**
 * Chart.js registration utility
 * Centralizes all Chart.js component registration
 */
export const registerChartComponents = (): void => {
  ChartLibrary.register(
    CategoryScale,
    LinearScale,
    Title,
    Tooltip,
    Legend,
    RadialLinearScale,
    BarElement,
    ArcElement,
    PointElement,
    LineElement,
    BarController,
    DoughnutController,
    ScatterController,
    PolarAreaController,
  );
};

/**
 * Default theme configuration for charts
 */
export const defaultChartTheme = {
  colors: {
    primary: "#3b82f6",
    secondary: "#8b5cf6", 
    success: "#10b981",
    warning: "#f59e0b",
    danger: "#ef4444",
    background: "#ffffff",
    foreground: "#1f2937",
  },
  
  fonts: {
    title: {
      size: 16,
      weight: "bold" as const,
    },
    body: {
      size: 14,
      weight: "normal" as const,
    },
    legend: {
      size: 12,
      weight: "normal" as const,
    },
  },
  
  spacing: {
    padding: {
      top: 10,
      bottom: 20,
      left: 10,
      right: 10,
    },
    tooltip: 8,
  },
  
  borders: {
    width: 1,
    radius: 4,
  },
} as const;

/**
 * Default CSS classes for chart components
 */
export const defaultChartClasses = {
  root: "relative w-full h-max flex flex-col items-center border border-border justify-center bg-white dark:bg-content1 p-6 shadow-md rounded-xl",
  canvas: "w-full h-[400px]",
  title: "text-lg font-semibold text-center mb-4",
  legend: "mt-4",
  tooltip: "bg-white p-2 rounded shadow-lg border text-sm",
} as const;