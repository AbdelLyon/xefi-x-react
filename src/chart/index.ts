import "@/index.css";

export { Chart } from "@/chart/Chart";
export type {
  ChartBaseProps,
  ChartClassNames,
  ChartJS,
  ChartOptions,
  ChartProps,
  ChartType,
  ChartTypeRegistry,
} from "@/chart/Chart";
