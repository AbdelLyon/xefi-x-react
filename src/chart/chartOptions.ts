import type { ChartOptions, ChartTypeRegistry, TooltipItem } from "chart.js";
import { defaultChartTheme } from "./chartConfig";

export type ChartType = keyof ChartTypeRegistry;

/**
 * Options for creating default chart configuration
 */
export interface DefaultChartOptionsConfig<T extends ChartType> {
  responsive?: boolean;
  maintainAspectRatio?: boolean;
  title?: string;
  showLegend?: boolean;
  showTooltip?: boolean;
  legendPosition?: "top" | "bottom" | "left" | "right";
  customTooltip?: (context: TooltipItem<T>) => string | string[] | undefined;
}

/**
 * Creates default chart options with theme integration
 */
export const createDefaultChartOptions = <T extends ChartType>(
  config: DefaultChartOptionsConfig<T> = {},
): ChartOptions<T> => {
  const {
    responsive = true,
    maintainAspectRatio = false,
    title,
    showLegend = true,
    showTooltip = true,
    legendPosition = "top",
    customTooltip,
  } = config;

  return {
    responsive,
    maintainAspectRatio,
    plugins: {
      title: title
        ? {
          display: true,
          text: title,
          font: defaultChartTheme.fonts.title,
          padding: defaultChartTheme.spacing.padding,
        }
        : undefined,

      legend: {
        display: showLegend,
        position: legendPosition,
        labels: {
          font: defaultChartTheme.fonts.legend,
        },
      },

      tooltip: showTooltip
        ? {
          enabled: true,
          backgroundColor: defaultChartTheme.colors.background,
          titleColor: defaultChartTheme.colors.foreground,
          bodyColor: "#4b5563",
          borderColor: "#e5e7eb",
          borderWidth: defaultChartTheme.borders.width,
          padding: defaultChartTheme.spacing.tooltip,
          cornerRadius: defaultChartTheme.borders.radius,
          bodyFont: defaultChartTheme.fonts.body,
          titleFont: defaultChartTheme.fonts.title,
          ...(customTooltip && {
            callbacks: {
              label: customTooltip,
            },
          }),
        }
        : { enabled: false },
    },
  } as ChartOptions<T>;
};

/**
 * Merges user options with default options
 */
export const mergeChartOptions = <T extends ChartType>(
  defaultOptions: ChartOptions<T>,
  userOptions?: ChartOptions<T>,
): ChartOptions<T> => {
  if (!userOptions) { return defaultOptions; }

  return {
    ...defaultOptions,
    ...userOptions,
    plugins: {
      ...(defaultOptions?.plugins || {}),
      ...(userOptions?.plugins || {}),
    },
  };
};