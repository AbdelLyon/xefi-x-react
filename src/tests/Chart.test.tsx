// Chart.test.tsx
import type { Mock } from "vitest";
import { describe, it, expect, vi, beforeEach } from "vitest";
import { render } from "@testing-library/react";
import { Chart } from "@/chart";
import type { JSX } from "react";

vi.mock(
  "react-chartjs-2",
  (): {
    Chart: Mock<
      ({
        options,
      }: {
        options: { plugins: { title: { text: string } } };
      }) => JSX.Element
    >;
  } => ({
    Chart: vi.fn(
      ({ options }): JSX.Element => (
        <div data-testid="mock-chart">{options?.plugins?.title?.text}</div>
      ),
    ),
  }),
);

describe("Composant Chart", (): void => {
  beforeEach((): void => {
    vi.clearAllMocks();
  });

  const mockBarData = {
    labels: ["Rouge", "Bleu", "Jaune"],
    datasets: [
      {
        label: "Jeu de données en barres",
        data: [10, 20, 30],
        backgroundColor: ["#FF6384", "#36A2EB", "#FFCE56"],
      },
    ],
  };

  const mockLineData = {
    labels: ["Jan", "Fév", "Mar"],
    datasets: [
      {
        label: "Jeu de données en ligne",
        data: [5, 10, 15],
        borderColor: "#FF6384",
        tension: 0.1,
      },
    ],
  };

  const mockDoughnutData = {
    labels: ["A", "B", "C"],
    datasets: [
      {
        data: [300, 50, 100],
        backgroundColor: ["#FF6384", "#36A2EB", "#FFCE56"],
      },
    ],
  };

  const mockPolarAreaData = {
    labels: ["Rouge", "Bleu", "Jaune"],
    datasets: [
      {
        data: [11, 16, 7],
        backgroundColor: ["#FF6384", "#4BC0C0", "#FFCE56"],
      },
    ],
  };

  describe("Snapshots", (): void => {
    it("devrait correspondre au snapshot avec les props par défaut", (): void => {
      const { container } = render(<Chart type="bar" data={mockBarData} />);
      expect(container).toMatchSnapshot();
    });

    it("devrait correspondre au snapshot avec un titre personnalisé", (): void => {
      const { container } = render(
        <Chart type="bar" data={mockBarData} title="Graphique de test" />,
      );
      expect(container).toMatchSnapshot();
    });

    it("devrait correspondre au snapshot pour un graphique en ligne", (): void => {
      const { container } = render(
        <Chart
          type="line"
          data={mockLineData}
          options={{
            scales: {
              y: {
                beginAtZero: true,
              },
            },
          }}
        />,
      );
      expect(container).toMatchSnapshot();
    });

    it("devrait correspondre au snapshot pour un graphique en donut", (): void => {
      const { container } = render(
        <Chart
          type="doughnut"
          data={mockDoughnutData}
          options={{
            color: "red",
          }}
        />,
      );
      expect(container).toMatchSnapshot();
    });

    it("devrait correspondre au snapshot pour un graphique en aire polaire", (): void => {
      const { container } = render(
        <Chart
          type="polarArea"
          data={mockPolarAreaData}
          options={{
            scales: {
              r: {
                beginAtZero: true,
              },
            },
          }}
        />,
      );
      expect(container).toMatchSnapshot();
    });
  });

  describe("Configuration des options", (): void => {
    it("devrait correctement configurer les options de légende", (): void => {
      const { container } = render(
        <Chart
          type="doughnut"
          data={mockDoughnutData}
          options={{
            plugins: {
              legend: {
                position: "right",
                labels: {
                  padding: 20,
                },
              },
            },
          }}
        />,
      );
      expect(container).toMatchSnapshot();
    });

    it("devrait fusionner correctement les options par défaut avec les options personnalisées", (): void => {
      const customOptions = {
        responsive: false,
        plugins: {
          legend: {
            position: "bottom" as const,
          },
        },
      };

      const { container } = render(
        <Chart type="bar" data={mockBarData} options={customOptions} />,
      );
      expect(container).toMatchSnapshot();
    });
  });
});
