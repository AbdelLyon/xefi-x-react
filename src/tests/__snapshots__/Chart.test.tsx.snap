// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Composant Chart > Configuration des options > devrait correctement configurer les options de légende 1`] = `
<div>
  <div
    class="relative w-full h-max flex flex-col items-center border border-border justify-center bg-white dark:bg-content1 p-6 shadow-md rounded-xl"
  >
    <div
      data-testid="mock-chart"
    />
  </div>
</div>
`;

exports[`Composant Chart > Configuration des options > devrait fusionner correctement les options par défaut avec les options personnalisées 1`] = `
<div>
  <div
    class="relative w-full h-max flex flex-col items-center border border-border justify-center bg-white dark:bg-content1 p-6 shadow-md rounded-xl"
  >
    <div
      data-testid="mock-chart"
    />
  </div>
</div>
`;

exports[`Composant Chart > Snapshots > devrait correspondre au snapshot avec les props par défaut 1`] = `
<div>
  <div
    class="relative w-full h-max flex flex-col items-center border border-border justify-center bg-white dark:bg-content1 p-6 shadow-md rounded-xl"
  >
    <div
      data-testid="mock-chart"
    />
  </div>
</div>
`;

exports[`Composant Chart > Snapshots > devrait correspondre au snapshot avec un titre personnalisé 1`] = `
<div>
  <div
    class="relative w-full h-max flex flex-col items-center border border-border justify-center bg-white dark:bg-content1 p-6 shadow-md rounded-xl"
  >
    <div
      data-testid="mock-chart"
    >
      Graphique de test
    </div>
  </div>
</div>
`;

exports[`Composant Chart > Snapshots > devrait correspondre au snapshot pour un graphique en aire polaire 1`] = `
<div>
  <div
    class="relative w-full h-max flex flex-col items-center border border-border justify-center bg-white dark:bg-content1 p-6 shadow-md rounded-xl"
  >
    <div
      data-testid="mock-chart"
    />
  </div>
</div>
`;

exports[`Composant Chart > Snapshots > devrait correspondre au snapshot pour un graphique en donut 1`] = `
<div>
  <div
    class="relative w-full h-max flex flex-col items-center border border-border justify-center bg-white dark:bg-content1 p-6 shadow-md rounded-xl"
  >
    <div
      data-testid="mock-chart"
    />
  </div>
</div>
`;

exports[`Composant Chart > Snapshots > devrait correspondre au snapshot pour un graphique en ligne 1`] = `
<div>
  <div
    class="relative w-full h-max flex flex-col items-center border border-border justify-center bg-white dark:bg-content1 p-6 shadow-md rounded-xl"
  >
    <div
      data-testid="mock-chart"
    />
  </div>
</div>
`;
